### PR TITLE
Extended ACL system and Group support

### PR DESCRIPTION
**Pull request by berggren**
Extended ACL system for more granular permissions
* Per user ACL on sketches and timelines
* New sharing UI
* Group support. Sketches and timelines can now have group ACLs
* Automatic group creation and management via SSO

Ref issue: #157 
Reviewer: @onager

### CODE REVIEW

File: timesketch/models/acl.py [hunk 1]
```diff
@@ -106,19 +125,53 @@ def all_with_acl(cls, user=None):
...
+                    cls.AccessControlEntry.user == None,
+                    cls.AccessControlEntry.group == None),
+                cls.AccessControlEntry.group_id.in_(
+                    [group.id for group in user.groups])),
             cls.AccessControlEntry.permission == u'read',
             cls.AccessControlEntry.parent)
 
-    def _get_ace(self, user, permission):
+    def _get_ace(self, permission, user=None, group=None, check_group=True):
         """Get the specific access control entry for the user and permission.
```
Comment (Onager) [inline timesketch/models/acl.py]: Args?

File: timesketch/models/acl.py [hunk 2]
```diff
@@ -157,28 +210,52 @@ def has_permission(self, user, permission):
...
+        if group and not self._get_ace(permission, group=group):
             self.acl.append(
-                self.AccessControlEntry(
-                    user=user, permission=permission))
+                self.AccessControlEntry(permission=permission, group=group))
             db_session.commit()
+            return
 
-    def revoke_permission(self, user, permission):
+        # Grant permission to a user.
```
Comment (Onager) [inline timesketch/models/acl.py]: Incorrect comment, and you might want a docstring here, too

File: timesketch/models/acl.py [hunk 3]
```diff
@@ -157,28 +210,52 @@ def has_permission(self, user, permission):
...
+        if not self._get_ace(permission, user=user, check_group=False):
+            self.acl.append(
+                self.AccessControlEntry(permission=permission, user=user))
+            db_session.commit()
+
+    def revoke_permission(self, permission, user=None, group=None):
         """Revoke permission to a user with the specific permission.
```
Comment (Onager) [inline timesketch/models/acl.py]: This is a little confusing - what do you mean "with the specific permission"?

Comment (Onager) [inline timesketch/models/acl.py]: Oh, and this also looks like it revokes more than just user permissions, but groups as well.

**Comment by Onager** [comment]
LGTM, couple of little docstring suggestions.

File: timesketch/models/acl.py (hunk 1 above)
Comment (berggren) [inline timesketch/models/acl.py]: Done.

File: timesketch/models/acl.py (hunk 3 above)
Comment (berggren) [inline timesketch/models/acl.py]: Yes, this was a bit confusing :) Clarified a bit.

**Pull request merged**